From: Rene Kita <mail@rkta.de>
Subject: Enable re-ordering of buffers in the stack menu

I'm not happy with this implementation.

My original version did not use 'Buffer *prev'. It would check if we are at the
bottom of the stack with

    if (CurrentMenu->select == SelectMenu.nitem - 2) /* Bottom of menu */
       return (MENU_NOTHING);

but then GCC would warn about a potential null-dereference at

    if (cur == Firstbuf)
       Firstbuf = cur->nextBuffer;
    else
       prevBuffer(Firstbuf, cur)->nextBuffer = next;

because prevBuffer could return NULL if there were no previous buffer (which
can't happen because of the previous check).

Given the choice between losing a useful warning or wasting CPU cycles getting
prev even if we don't need it, I decided to do the unnecessary computation
because it's no big deal nowadays. Now I feel dirty like a whore.

---
 doc-de/MANUAL.html |   45 ++----------------------------
 doc/MANUAL.html    |   47 ++++---------------------------
 menu.c             |   78 ++++++++++++++++++++++++++++++++++++++++++++++++++++-
 3 files changed, 87 insertions(+), 83 deletions(-)
--- a/menu.c
+++ b/menu.c
@@ -242,6 +242,8 @@ static int SelectV = 0;
 static void initSelectMenu(void);
 static void smChBuf(void);
 static int smDelBuf(char c);
+static int smMovDn(char c);
+static int smMovUp(char c);
 
 /* --- SelectMenu (END) --- */
 
@@ -1387,7 +1389,8 @@ initSelectMenu(void)
     Str str;
     char **label;
     char *p;
-    static char *comment = " SPC for select / D for delete buffer ";
+    static const char *comment =
+	    " SPC: select | D: delete | </>: move down/up ";
 
     SelectV = -1;
     for (i = 0, buf = Firstbuf; buf != NULL; i++, buf = buf->nextBuffer) {
@@ -1442,6 +1445,8 @@ initSelectMenu(void)
     SelectMenu.cursorX = Currentbuf->cursorX + Currentbuf->rootX;
     SelectMenu.cursorY = Currentbuf->cursorY + Currentbuf->rootY;
     SelectMenu.keymap['D'] = smDelBuf;
+    SelectMenu.keymap['>'] = smMovUp;
+    SelectMenu.keymap['<'] = smMovDn;
     SelectMenu.item[nitem].type = MENU_NOP;
 }
 
@@ -1506,6 +1511,77 @@ smDelBuf(char c)
     return (MENU_NOTHING);
 }
 
+static int
+smMovUp(char c)
+{
+    Buffer *cur, *prev, *pprev;
+    int i, mselect, x, y;
+
+    if (!CurrentMenu->select) /* Top of menu */
+	return MENU_NOTHING;
+    for (i = 0, cur = Firstbuf; i < CurrentMenu->select;
+	 i++, cur = cur->nextBuffer) ;
+    prev = prevBuffer(Firstbuf, cur);
+    if ((pprev = prevBuffer(Firstbuf, prev)))
+	pprev->nextBuffer = cur;
+    prev->nextBuffer = cur->nextBuffer;
+    cur->nextBuffer = prev;
+    if (prev == Firstbuf)
+	Firstbuf = cur;
+
+    x = CurrentMenu->x;
+    y = CurrentMenu->y;
+    mselect = CurrentMenu->select;
+
+    initSelectMenu();
+    CurrentMenu->x = x;
+    CurrentMenu->y = y;
+    geom_menu(CurrentMenu, 0, 0, 0);
+    CurrentMenu->select = --mselect;
+    displayBuffer(Currentbuf, B_FORCE_REDRAW);
+    draw_all_menu(CurrentMenu);
+    select_menu(CurrentMenu, CurrentMenu->select);
+    return MENU_NOTHING;
+}
+
+static int
+smMovDn(char c)
+{
+    Buffer *cur, *next, *prev;
+    int i, mselect, x, y;
+
+    /* Get selected buffer */
+    for (i = 0, cur = Firstbuf; i < CurrentMenu->select;
+	 i++, cur = cur->nextBuffer) ;
+
+    if (!cur->nextBuffer) /* Bottom of menu */
+	return MENU_NOTHING;
+
+    prev = prevBuffer(Firstbuf, cur);
+    next = cur->nextBuffer;
+
+    if (cur == Firstbuf)
+	Firstbuf = cur->nextBuffer;
+    else
+	prev->nextBuffer = next;
+    cur->nextBuffer = next->nextBuffer;
+    next->nextBuffer = cur;
+
+    x = CurrentMenu->x;
+    y = CurrentMenu->y;
+    mselect = CurrentMenu->select;
+
+    initSelectMenu();
+    CurrentMenu->x = x;
+    CurrentMenu->y = y;
+    geom_menu(CurrentMenu, 0, 0, 0);
+    CurrentMenu->select = ++mselect;
+    displayBuffer(Currentbuf, B_FORCE_REDRAW);
+    draw_all_menu(CurrentMenu);
+    select_menu(CurrentMenu, CurrentMenu->select);
+    return MENU_NOTHING;
+}
+
 /* --- SelectMenu (END) --- */
 
 /* --- SelTabMenu --- */
--- a/doc/MANUAL.html
+++ b/doc/MANUAL.html
@@ -1931,47 +1931,12 @@ mh 2016-06-12 Interrupt is an alias to S
 	<h3>Menu Selection Mode</h3>
       </td>
     </tr>
-      <tr>
-     <td>
-     BUF:NEXT</td>
-    <td>
-     j, C-n, DOWN
-    </td>
-    <td>
-     Next item
-    </td>
-   </tr>
-      <tr>
-     <td>BUF:PREV
-     </td>
-    <td>
-     k, C-p, UP
-    </td>
-    <td>
-     Previous item
-    </td>
-   </tr>
-      <tr>
-     <td>
-     BUF:DELETE</td>
-    <td>
-     D
-    </td>
-    <td>
-     Delete item
-    </td>
-   </tr>
-      <tr>
-     <td>
-       BUF:GO
-     </td>
-    <td>
-     SPC, RET, RIGHT
-    </td>
-    <td>
-     Select item
-    </td>
-   </tr>
+    <tr><td>N/A</td><td>j, C-n, DOWN</td><td>Next item</td></tr>
+    <tr><td>N/A</td><td>k, C-p, UP</td><td>Previous item</td></tr>
+    <tr><td>N/A</td><td>D</td><td>Delete item</td></tr>
+    <tr><td>N/A</td><td>SPC, RET, RIGHT</td><td>Select item</td></tr>
+    <tr><td>N/A</td><td>&lt;</td><td>Move down</td></tr>
+    <tr><td>N/A</td><td>&gt;</td><td>Move up</td></tr>
 <!--  </table>-->
     <tr>
       <td colspan="3">
--- a/doc-de/MANUAL.html
+++ b/doc-de/MANUAL.html
@@ -1890,47 +1890,10 @@ mh 2016-06-12 Interrupt is an alias to S
 	<h3>Menü-Auswahl-Modus</h3>
       </td>
     </tr>
-      <tr>
-     <td>
-     BUF:NEXT</td>
-    <td>
-     j, C-n, DOWN
-    </td>
-    <td>
-     Nächster Eintrag
-    </td>
-   </tr>
-      <tr>
-     <td>BUF:PREV
-     </td>
-    <td>
-     k, C-p, UP
-    </td>
-    <td>
-     Vorheriger Eintrag
-    </td>
-   </tr>
-      <tr>
-     <td>
-     BUF:DELETE</td>
-    <td>
-     D
-    </td>
-    <td>
-     Lösche Eintrag
-    </td>
-   </tr>
-      <tr>
-     <td>
-       BUF:GO
-     </td>
-    <td>
-     SPC, RET, RIGHT
-    </td>
-    <td>
-     Wähle Eintrag aus
-    </td>
-   </tr>
+    <tr><td>N/A</td><td>j, C-n, DOWN</td><td>Nächster Eintrag</td></tr>
+    <tr><td>N/A</td><td>k, C-p, UP</td><td>Vorheriger Eintrag</td></tr>
+    <tr><td>N/A</td><td>D</td><td>Lösche Eintrag</td></tr>
+    <tr><td>N/A</td><td>SPC, RET, RIGHT</td><td>Wähle Eintrag aus</td></tr>
 <!--  </table>-->
     <tr>
       <td colspan="3">
